subprojects/dependency-locking/src/main/java/org/gradle/dependency/locking/DependencyLockingPlugin.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.dependency.locking;
+
+import org.gradle.api.Action;
+import org.gradle.api.Incubating;
+import org.gradle.api.Plugin;
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.internal.artifacts.dsl.dependencies.DependencyFactory;
+import org.gradle.api.internal.project.ProjectInternal;
+import org.gradle.api.logging.Logger;
+import org.gradle.api.logging.Logging;
+
+import javax.inject.Inject;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+@Incubating
+public class DependencyLockingPlugin implements Plugin<ProjectInternal> {
+
+    private static final Logger LOGGER = Logging.getLogger(DependencyLockingPlugin.class);
+
+    private final AtomicBoolean mustWrite = new AtomicBoolean(false);
+    private LockFileReaderWriter lockFileReaderWriter;
+    private final DependencyFactory dependencyFactory;
+
+    @Inject
+    public DependencyLockingPlugin(DependencyFactory dependencyFactory) {
+        this.dependencyFactory = dependencyFactory;
+    }
+
+    @Override
+    public void apply(final ProjectInternal project) {
+        LOGGER.warn("Applying dependency-locking plugin");
+        lockFileReaderWriter = new LockFileReaderWriter(project);
+
+        DependencyLockTask saveDependencyLocks = project.getTasks().create("saveDependencyLocks", DependencyLockTask.class);
```

Comment:
As I commented on the doc already, I think we should explore the plugin without tasks and just configuration.

If we have `saveDependencyLocks`  as a task, the build author or other plugin authors need to make sure that the task does not run too early...

Reply:
I think having the task is valuable: new lock files generation only happens when explicitly requested. As for the ordering issue, the goal is to make it irrelevant.